File: src/main/java/com/warren/loadouts/InventoryLoadout.java
package com.warren.loadouts;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

import org.powbot.api.rt4.Inventory;

public class InventoryLoadout implements ItemLoadout {
  private List<ItemEntry> required;

  public InventoryLoadout(List<ItemEntry> entries) {
    this.required = entries;
  }
  
  public InventoryLoadout require(ItemEntry entry) {
    required.add(entry);
    return this;
  }
  
  public static InventoryLoadout of(ItemEntry... entries) {
    return new InventoryLoadout(entries == null ? List.of() : Arrays.asList(entries));
  }

  @Override
  public Iterator<ItemEntry> iterator() {
    return required.iterator();
  }
}

File: src/main/kotlin/com/warren/loadouts/InventoryLoadout.kt
package com.warren.loadouts

import org.powbot.api.rt4.Item

class InventoryLoadout internal constructor(
    override val entries: List<ItemEntry>
) : ItemLoadout {
    override fun toString(): String = "InventoryLoadout(entries=$entries)"

    companion object {
        @JvmStatic
        fun builder(): InventoryLoadoutBuilder = InventoryLoadoutBuilder()
    }
}

class InventoryLoadoutBuilder {
    private val entries = mutableListOf<ItemEntry>()

    fun entry(block: ItemEntryBuilder.() -> Unit) {
        entries += ItemEntryBuilder().apply(block).build()
    }

    fun item(vararg names: String, block: ItemEntryBuilder.() -> Unit = {}) {
        entries += ItemEntryBuilder().apply {
            names(*names)
            block()
        }.build()
    }

    fun build(): InventoryLoadout = InventoryLoadout(entries.toList())
}
